palace/fem/qfunctions/apply_qf.h

```diff
@@ -0,0 +1,191 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef PALACE_LIBCEED_APPLY_QF_H
+#define PALACE_LIBCEED_APPLY_QF_H
+
+// libCEED QFunctions for application of a generic operator with assembled quadrature data.
+// in[0] is (symmetric) quadrature data, shape [ncomp=vdim*(vdim+1)/2, Q]
+// in[1] is active vector, shape [ncomp=vdim, Q]
+// out[0] is active vector, shape [ncomp=vdim, Q]
+
+// For pairwise apply functions, the inputs and outputs come in pairs and the quadrature
+// data is arranged to be applied with the first vdim*(vdim+1)/2 components for the first
+// input/output and the remainder for the second.
+
+CEED_QFUNCTION(f_apply_1)(void *, CeedInt Q, const CeedScalar *const *in,
+                          CeedScalar *const *out)
+{
+  const CeedScalar *__restrict__ qd = in[0], *__restrict__ u = in[1];
```

Comment:
`__restrict__` tells the compiler that pointers are not aliasing so that, for example, it is safe to vectorize certain loops or other operations. This was needed to get a few things to vectorize (mainly with the material property coefficients stored in `ctx`). See also https://developer.nvidia.com/blog/cuda-pro-tip-optimize-pointer-aliasing/, it can be particularly relevant for GPUs.